Implementation of Dynamic Window Pure Pursuit (DWPP)

Decwest opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5524 |
| Primary OS tested on | Ubuntu 24.04 (ROS2 rolling) |
| Robotic platform tested on | gazebo simulation of turtlebot3_waffle |
| Does this PR contain AI generated software? | No |
| Was this PR description generated by AI software? | No |

---

## Description of contribution in a few bullet points

<!--
* I added this neat new feature
* Also fixed a typo in a parameter name in nav2_costmap_2d
-->

- Implement DWPP as an extension to the current RPP implementation

## Description of documentation updates required from your changes

<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

- Added new parameter, so need to add that to default configs and documentation page

**Added Parameters**

- **desired_angular_vel** (double): maximum angular velocity (rad/s)
- **max_linear_accel** (double): maximum linear acceleration (m/s²)
-  **use_dynamic_window** (bool): enable DWPP
- **velocity_feedback** (string, no need to change) : How the current velocity is obtained during dynamic window computation.
  - `"OPEN_LOOP"`: Uses the last commanded velocity (recommended)
  - `"CLOSED_LOOP"`: Uses odometry velocity (may hinder proper acceleration/deceleration)

## Description of how this change was tested

<!--
* I wrote unit tests that cover 90%+ of changes and extensively tested on my physical robot platform in production for 1 week
* I wrote unit tests and tested in simulation for 10 minutes
* Performed linting validation using pre-commit run --all or colcon test
-->

- Built a simulation environment that allows comparison between different methods, and verified the implementation and effectiveness of DWPP
  - Docker environment: https://github.com/Decwest/dwpp_test_environment/tree/feature/nav2_integration
  - Nav2 parameter file used during simulation: https://github.com/Decwest/dwpp_test_simulation/blob/b90e1c0e43885f1146b16cff64f4358b5907b110/params/test_params.yaml

**Simulation video**

https://github.com/user-attachments/assets/3e065b1f-d1ad-4194-b6d6-271ae577600e

---

## Future work that may be required in bullet points

<!--
* I think there might be some optimizations to be made from STL vector
* I see a lot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

- Discussion and modification regarding the renaming of Regulated Pure Pursuit
- Documentation update

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

Decwest commented:
@SteveMacenski 
Thank you for reviewing again, and sorry for the delay in my response. 
I’ll address all of your comments one by one and will mention you once everything has been updated. 
I’ll also work on improving the code readability.

SteveMacenski commented:
Let me know when I should take a look again! Note that you should probably rebase and/or pull in main so that CI can pass :-)

SteveMacenski commented:
Please fix the merge conflict 

Sorry for the delay. ROSCon + vacation took me out of the office for some time and takes a moment to get up to speed on all the PRs again that have piled up

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/parameter_handler.cpp:
```diff
@@ -37,7 +38,13 @@ ParameterHandler::ParameterHandler(
   logger_ = logger;
 
   declare_parameter_if_not_declared(
-    node, plugin_name_ + ".desired_linear_vel", rclcpp::ParameterValue(0.5));
```
We should have a migration log here. If `desired_linear_vel` is set, log a warning that it has been changed to `max_linear_vel` and that they should migrate the use.

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/parameter_handler.cpp:
```diff
@@ -106,9 +119,16 @@ ParameterHandler::ParameterHandler(
+  declare_parameter_if_not_declared(
+      node, plugin_name_ + ".use_dynamic_window", rclcpp::ParameterValue(true));
+  declare_parameter_if_not_declared(
+      node, plugin_name_ + ".velocity_feedback", rclcpp::ParameterValue(std::string("OPEN_LOOP")));
```
Why is this adding a new feature? Please remove and open another PR to add.

SteveMacenski replied:
It is much easier for maintainers to review 2x decoupled smaller PRs than 1 larger PR

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -164,6 +165,228 @@ double calculateCurvature(geometry_msgs::msg::Point lookahead_point)
   }
 }
 
+void RegulatedPurePursuitController::computeDynamicWindow(
```
I think this should have a struct return type that is the dynamic window limits. This implicit return thing is a bit awkward

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -164,6 +165,228 @@ double calculateCurvature(geometry_msgs::msg::Point lookahead_point)
+
+  const double & dt = control_duration_;
+
+  constexpr double Eps = 1e-2;
```
```suggestion
  constexpr double Eps = 1e-3;
```

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -164,6 +165,228 @@ double calculateCurvature(geometry_msgs::msg::Point lookahead_point)
+  double & dynamic_window_min_linear_vel)
+{
+  // Extract the portion of the dynamic window that lies within the range [0, regulated_linear_vel]
+  double dynamic_window_max_linear_vel_temp;
```
Don't name things "temp", that's bad SE practice. Name it meaningfully

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -164,6 +165,228 @@ double calculateCurvature(geometry_msgs::msg::Point lookahead_point)
+                 dynamic_window_min_angular_vel);
+}
+
+void RegulatedPurePursuitController::applyRegulationToDynamicWindow(
```
This seems like it can be a utils function as well; it doesn't use class members

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -164,6 +165,228 @@ double calculateCurvature(geometry_msgs::msg::Point lookahead_point)
+        optimal_angular_vel = dynamic_window_max_angular_vel;
+      }
+    }
+    return;
```
```suggestion
```

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -271,7 +494,49 @@ geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocity
+      // considering velocity and acceleration constraints (DWPP)
+      const double regulated_linear_vel = linear_vel;
+      geometry_msgs::msg::Twist current_speed;
+      if (params_->velocity_feedback == "CLOSED_LOOP") {
```
Shouldn't be be made of use more generally if we're enabling open vs closed loop? This is not the only place speed is utilized

Decwest commented on nav2_regulated_pure_pursuit_controller/src/parameter_handler.cpp:
```diff
@@ -88,7 +95,13 @@ ParameterHandler::ParameterHandler(
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".max_linear_decel", rclcpp::ParameterValue(0.5));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".max_angular_accel", rclcpp::ParameterValue(1.0));
```
I set these same as current RPP's default value.

Decwest commented on nav2_regulated_pure_pursuit_controller/src/parameter_handler.cpp:
```diff
@@ -37,7 +38,13 @@ ParameterHandler::ParameterHandler(
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".min_linear_vel", rclcpp::ParameterValue(0.0));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".max_angular_vel", rclcpp::ParameterValue(1.0));
```
I set these same as velocity smoother's default values

Decwest commented:
@SteveMacenski 
Thank you very much for taking the time to provide such a careful review despite your busy schedule.
I will address all of your comments once again.

I have one issue that I would like to ask for your advice on.
At the moment, I am developing in a Rolling Docker environment based on osrf/ros:rolling-desktop-full:
https://github.com/Decwest/dwpp_test_environment/blob/feature/nav2_integration/docker/Dockerfile.rolling

However, after syncing Nav2 to the latest version today and building and running it in this environment, I encountered a segmentation fault, and Nav2 no longer starts.
If you happen to have any insight into what might be causing this, I would greatly appreciate your advice.

Below is a detailed description of the situation.
- **Log obtained when running `controller_server` in debug mode**
```shell
ROLLING ubuntu:~/ros2_ws$ ros2 run nav2_controller controller_server \
  --ros-args --log-level debug \
  -r __ns:=/ \
  -p use_sim_time=False
[DEBUG] [1764069038.870090908] [rclcpp]: signal handler installed
[DEBUG] [1764069038.870102907] [rclcpp]: deferred_signal_handler(): waiting for SIGINT/SIGTERM or uninstall
[DEBUG] [1764069038.870473289] [rcl]: Initializing node 'controller_server' in namespace ''
[DEBUG] [1764069038.870496899] [rcl]: Using domain ID of '0'
[ros2run]: Segmentation fault
```
From this, it seems that the process runs at least up to `rclcpp::init()` and `rcl_init()`, but it crashes immediately afterward. In addition to controller_server, planner_server, behavior_server, smoother_server, route_server, and opennav_docking all crashed as well. bt_navigator, waypoint_follower, collision_monitor, velocity_smoother, and lifecycle_manager remained alive. No other error messages were printed.

- **Clean build**
I deleted the build, install, and log directories and rebuilt everything, but the issue persisted.

- **Recreating the environment**
I cloned the repository from scratch on another PC and rebuilt the Docker environment, but the same problem occurred. Because of this, I believe it is unlikely that any local files are causing the issue.

- **Updating binaries**
I ran sudo apt update and upgrade and then performed a clean build, but the same issue occurred.

- **Downgrading the Nav2 version**
I reverted the Nav2 commit to an earlier one (10/26, https://github.com/Decwest/navigation2/commit/d536d33be89a5291f72e08b6bbaa7f0f374a57e2
) and performed a clean build, but the issue persisted.

- **Reverting the Docker image to a previous version**
I attempted this, but I could not find older tags for osrf/ros:rolling-desktop-full. When I ran Docker build today, the image was cloned from scratch, so it is certain that the base image has been updated since the last time the environment was working. My suspicion is that this update might be related to the issue.

If you have any ideas or suggestions, I would be grateful to hear them.
Thank you very much for your time and assistance.

Decwest commented:
Thank you very much for providing such helpful information.

After trying a very, very large number of things, I found that building BehaviorTree.CPP from source resolved the issue described above.
https://github.com/BehaviorTree/BehaviorTree.CPP

GDB did not produce any readable clues, and neither the Rolling version nor the Nav2 version had anything to do with the issue.
However, the presence or absence of the issue clearly changed depending on whether BehaviorTree.CPP was built from source, so that is certainly the cause. The only thing I still do not understand is why a source build suddenly became necessary despite everything having worked fine until now.

Anyway, the clue that led me to the solution came from the fact that the nav2_docker you introduced builds BehaviorTree.CPP from source. I would like to express my sincere gratitude once again.

I will continue addressing the comments you provided.

SteveMacenski commented:
```
git rebase HEAD~41 --signoff
git push --force-with-lease origin feature/implement_dwpp
```

Doesn't work?  Maybe it would be best to get your work, squash it into a single commit that you sign off on and force push back? This is an important thing we need for contributors to verify that they have permission to contribute their work to the project.  

Sorry for the delay, I'm not sure why this didn't come up in my notification tray.

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/README.md:
```diff
@@ -165,3 +187,7 @@ To tune to get Pure Pursuit behaviors, set all boolean parameters to false and m
+
+When `use_dynamic_window` is set to True, the velocity, acceleration, and deceleration limits are enforced during the velocity command computation.
+Note that the velocity smoother clips the velocity commands output by this controller based on its own velocity and acceleration constraints before publishing cmd_vel.
+Therefore, the velocity smoother’s `max_velocity`, `min_velocity`, `max_accel`, and `max_decel` parameters must be consistent with this controller’s corresponding velocity, acceleration, and deceleration settings.
```
```suggestion
Therefore, the velocity smoother’s `max_velocity`, `min_velocity`, `max_accel`, and `max_decel` parameters must be consistent or greater than this controller’s corresponding velocity, acceleration, and deceleration settings.
```

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -271,7 +272,46 @@ geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocity
+      dynamic_window_pure_pursuit::DynamicWindowBounds dynamic_window =
+        dynamic_window_pure_pursuit::computeDynamicWindow(
+        current_speed, max_linear_vel, min_linear_vel, max_angular_vel, min_angular_vel,
+        max_linear_accel, max_linear_decel, max_angular_accel, max_angular_decel,
```
```suggestion
          current_speed, max_linear_vel, min_linear_vel, max_angular_vel, min_angular_vel,
          max_linear_accel, max_linear_decel, max_angular_accel, max_angular_decel,
```

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:
```diff
@@ -271,7 +272,46 @@ geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocity
+        dynamic_window_pure_pursuit::computeOptimalVelocityWithinDynamicWindow(
+        dynamic_window,
+        regulation_curvature,
+        x_vel_sign);
```
Could we not simply reduce this to 

```
dynamic_window_pure_pursuit::computeDynamicWindowVelocities(...)
```

Given all the inputs required in one API call? That would then call these 3 methods in order and/or consolidate a bit. It would make the application code here more streamline. Also would make the `const double & max_linear_vel = params_->max_linear_vel;` lines not as important since you could just do `params_->max_linear_vel` inline the API call

Decwest replied:
I implemented like this
https://github.com/ros-navigation/navigation2/blob/7d9fbeeb6da688594ed09e74cb6c77daa4ac8270/nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/dynamic_window_pure_pursuit_functions.hpp#L300

SteveMacenski commented on nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/dynamic_window_pure_pursuit_functions.hpp:
```diff
@@ -0,0 +1,286 @@
+};
+
+/**
+ * @brief Compute the dynamic window (feasible velocity bounds) based on the current speed and the given velocity and acceleration constraints.
```
This is definitely outside of the linting bounsd
